### ISSUE TITLE
Flow type inference for commitMutation and useMutation

### ISSUE THREAD

**Issue by leoasis** [facebook/relay#3995]
We noticed that the type signature of `commitMutation` and `useMutation` don't take advantage of the artifacts' exported type in order to infer data and variables.

Looking at [fetchQuery](https://github.com/facebook/relay/blob/396834117d30a6100abf2700979beca86340f143/packages/relay-runtime/query/fetchQuery.js#L115), [useFragment](https://github.com/facebook/relay/blob/396834117d30a6100abf2700979beca86340f143/packages/react-relay/relay-hooks/useFragment.js#L30), and [useLazyLoadQuery](https://github.com/facebook/relay/blob/396834117d30a6100abf2700979beca86340f143/packages/react-relay/relay-hooks/useLazyLoadQuery.js#L34) as an example, we can see how they use the `Query` and `Fragment` types.

However, [commitMutation](https://github.com/facebook/relay/blob/396834117d30a6100abf2700979beca86340f143/packages/relay-runtime/mutations/commitMutation.js#L37) and [useMutation](https://github.com/facebook/relay/blob/396834117d30a6100abf2700979beca86340f143/packages/react-relay/relay-hooks/useMutation.js#L54) still seem to be using the `GraphQLTaggedNode` type, even while the compiler generates artifacts that are of type `Mutation`.

Because of that, these 2 functions don't infer the types from the artifacts passed, don't validate that they're getting a mutation passed in, and rely on having to manually annotate the Mutation:

```js
const [commit] = useMutation<MyMutation>(myMutation, ...);
```

instead of just 

```js
const [commit] = useMutation(myMutation, ...);
```

Would you be willing to accept a PR that modifies these signatures so that they leverage the compiler generated types in the same way as `Fragment` and `Query`?

**Comment by captbaritone** [comment]
Thanks for writing this up! I think it's a great improvement, the only impediment is that we will need to find someone internally to work on the rollout (codemod and supress any Flow errors it uncovers). I'm asking around to see if anyone is interested.

Would you roll this out at Twitter by manually updating call sites, or would you use a codemod?

**Comment by leoasis** [comment]
Kinda, since we're not yet in a Flow version that has the Relay integration enabled, we're faking this with an autofixable ESLint rule (made by @lpalmes) that automatically imports the type and explicitly casts the `graphql` tags to the types exported by their corresponding artifact.

In our case, we would modify the rule so that it autofixes old usages.

But yeah, kind of like using a codemod.

The interesting thing is that our ESLint rule has a lot of added complexity that we cannot remove because of this case, as otherwise the logic is quite simple.

**Comment by captbaritone** [comment]
Would you be able to contribute a version of that rule which just applies the fix? Would be a super useful tool for everyone navigating the API change (us included).

**Comment by leoasis** [comment]
Oh, Just to clarify, and to make sure it would be useful to you, what our lint rule does is take things like:

```js
// @flow strict
useLazyLoadQuery(graphql` query AppQuery { viewer { id } } `)
```

and convert it to 

```js
import typeof AppQuery from 'src/__generated__/AppQuery.graphql';

useLazyLoadQuery((graphql`query AppQuery { viewer { id } } `: AppQuery), {})
```

it also handles the case where you assign your artifact to a variable:

```js
const simpleQuery = graphql`
  query AppQuery { viewer { id } }
`
```

converted to 

```js
import typeof AppQuery from 'src/__generated__/AppQuery.graphql';

const simpleQuery: AppQuery = graphql`
  query AppQuery { viewer { id } }
`
```

And to handle the `useMutation` case, since it's different, we actually based this off of your existing ESLint plugin: https://github.com/relayjs/eslint-plugin-relay/blob/main/src/rule-generated-flow-types.js#L570-L583

We actually did use a codemod to strip the type annotations out of all Relay API calls when we upgraded to Relay 13, but I think with this rule handling the `useMutation` case and change it so that it removes the type annotation, it should work.

Would this be something useful to you? I would assume only stripping out the type annotations would be what you need to do.

**Comment by tomgasson** [comment]
We're also on a pre-`relay_integration` version of flow (pre-`indexed-types`, and `types-first` in fact 😕) at Atlassian for Jira, but soon to migrate to TypeScript to line up with our other products.

We're having a lot of difficulty with this. We've currently got a mix of where we've backwards patched functions to their pre flow `relay_integration` types, and a handful of very verbose calls (`use*<Thing$variables, Thing$data>(graphql'')` style).

I think the way forward for non-`relay_integration` types is the format @leoasis is using, since it's annotating the source rather than relying on inference.

The TypeScript types should ideally go the same way to shore things up. You can still have fragment confusion there
```
const adata = useFragment(graphql`fragment A {}`, akey) // A$data

// Whoops! Passed the wrong key, incorrectly inferred the wrong data 
const bdata = useFragment(graphql`fragment B {}`, akey) // A$data. 
```

Would really appreciate if you shared the lint rule - it would be very useful. I was about to write it myself

**Comment by maraisr** [comment]
@n1ru4l i vaguely remember you pocing out a way to use template interpolation types to verify the `a$key` is accurate to its `graphql` template? Was that you or @sibelius??

**Comment by leoasis** [comment]
> Would really appreciate if you shared the lint rule - it would be very useful. I was about to write it myself

Hello @tomgasson sorry for not replying here earlier. Are you still in need to publish this eslint rule? We'd be happy to share it if it helps folks.

@captbaritone let us know if this is also helpful to you at Meta. Would be great to help in whatever way so we could move mutations to behave the same way as queries in terms of type inference, to simplify our tooling.

**Comment by tomgasson** [comment]
@leoasis: Don't need it anymore, we moved to typescript 😁, are back in the verbose format 😢 . Still having more difficulty there than desired in TS - looking for what devX improvements we could do there

**Comment by alunyov** [comment]
Should be closed with 603ce546741aabaa7adf4804f7c825d9ef69027c and 051e4694c20b6fa5a5a7a6d62e52f6a6a9919349
